staging/src/k8s.io/code-generator/cmd/api-linter/linters/api-linter.go

```diff
@@ -0,0 +1,245 @@
+
+func getAPITagValue(comments []string) []string {
+	return types.ExtractCommentTags("+", comments)[tagName]
+}
+
+func hasAPITagValue(comments []string, value string) bool {
+	tagValues := getAPITagValue(comments)
+	for _, val := range tagValues {
+		if val == value {
+			return true
+		}
+	}
+	return false
+}
+
+// hasOptionalTag returns true if the member has +optional in its comments or
+// omitempty in its json tags.
+func hasOptionalTag(m *types.Member) bool {
+	hasOptionalCommentTag := types.ExtractCommentTags(
+		"+", m.CommentLines)[tagOptional] != nil
+	hasOptionalJSONTag := strings.Contains(
+		reflect.StructTag(m.Tags).Get("json"), "omitempty")
+	return hasOptionalCommentTag || hasOptionalJSONTag
+}
+
+// NameSystems returns the name system used by the generators in this package.
+func NameSystems() namer.NameSystems {
+	return namer.NameSystems{
+		"raw": namer.NewRawNamer("", nil),
+	}
+}
+
+// DefaultNameSystem returns the default name system for ordering the types to be
+// processed by the generators in this package.
+func DefaultNameSystem() string {
+	return "raw"
+}
+
+// Packages returns a API linter that validates API conventions and generates a dummy package.
+func Packages(context *generator.Context, arguments *args.GeneratorArgs) generator.Packages {
+	return generator.Packages{
+		&generator.DefaultPackage{
+			PackageName: "dummy",
+			PackagePath: "k8s.io/kubernetes/staging/src/k8s.io/code-generator/cmd/api-linter/dummy",
+			GeneratorFunc: func(c *generator.Context) (generators []generator.Generator) {
+				filename := ""
+				if customArgs, ok := arguments.CustomArgs.(*CustomArgs); ok {
+					filename = customArgs.WhitelistFilename
+				}
+				apiValidators := []APIValidator{
+					optionalFieldAPIConvention{},
+				}
+				return []generator.Generator{newAPILinter(arguments.OutputFileBaseName, filename, apiValidators)}
+			},
+			FilterFunc: func(c *generator.Context, t *types.Type) bool {
+				// There is a conflict between this codegen and codecgen, we should avoid types generated for codecgen
+				if strings.HasPrefix(t.Name.Name, "codecSelfer") {
+					return false
+				}
+				pkg := context.Universe.Package(t.Name.Package)
+				if hasAPITagValue(pkg.Comments, tagValueTrue) {
+					return !hasAPITagValue(t.CommentLines, tagValueFalse)
+				}
+				return hasAPITagValue(t.CommentLines, tagValueTrue)
+			},
+		},
+	}
+}
+
+func (l *apiLinter) Filter(c *generator.Context, t *types.Type) bool {
+	// There is a conflict between this codegen and codecgen, we should avoid types generated for codecgen
+	return !strings.HasPrefix(t.Name.Name, "codecSelfer")
+}
+
+func (l *apiLinter) Imports(c *generator.Context) []string {
+	return []string{}
+}
+
+// apiLinter validates API spec against API conventions.
+type apiLinter struct {
+	generator.DefaultGen
+	validators        []APIValidator
+	violations        []apiViolation
+	whitelist         map[apiViolation]bool
+	whitelistFilename string
+}
+
+type apiViolation struct {
```

Comment:
Can you document any rules about setting these?  Like what is a suitable linterName or violationID?

Also, do we have access to file and line number where the type is declared?  This makes tracking down the failure a bit easier...  e.g., here's one from my conformance regtest.

```
ERROR at test/e2e/scheduling/predicates.go:245:2: Using It() with manual [Conformance] tag is no longer allowed.  Use framework.ConformanceIt() instead.
```

Reply:
Updated. Unfortunately we don't have access to file and line number in gengo.

Reply:
Am I looking at the wrong file?  I don't see the update.